### PR TITLE
sctp/data channel refactoring

### PR DESCRIPTION
**Pull request by algesten**
This tidies up the init sequence for data channels in anticipation of direct api needs.

The correct sequence is:

```
self.rtc.chan.new_channel(&config);
self.rtc.chan.confirm(id, config);
```

There is a new `ChannelConfig` object to configure all the options of the data channel.

### CODE REVIEW

File: src/sctp/mod.rs [hunk 1]
```diff
@@ -72,22 +72,79 @@ impl RtcSctpState {
...
+    /// If we have sent DcepOpen and are waiting for the ack.
     AwaitDcepAck,
+    /// Stream is open, ready to send data.
     Open,
+    /// If some error occurs.
     Closed,
 }
 
+#[derive(Debug, Default)]
+pub struct ChannelConfig {
```
Comment (xnorpx) [inline src/sctp/mod.rs:127]: should this be exported in lib?

File: src/channel.rs [hunk 2]
```diff
@@ -52,3 +64,169 @@ impl fmt::Debug for ChannelData {
...
+        let id = ChannelId(self.allocations.len());
+
+        // For out-of-band negotiated, the id is already set.
+        let sctp_stream_id = config.negotiated;
+        if let Some(sctp_stream_id) = sctp_stream_id {
+            let exists = self
+                .allocations
+                .iter()
+                .any(|a| a.sctp_stream_id == Some(sctp_stream_id));
+            assert!(exists, "sctp_stream_id ({}) exists already", sctp_stream_id);
```
Comment (xnorpx) [inline src/channel.rs:96]: should this be !exists?

File: src/sctp/mod.rs (hunk 1 above)
Comment (algesten) [reply src/sctp/mod.rs:127]: Yes. Once we got an API that uses it.

File: src/channel.rs (hunk 2 above)
Comment (algesten) [reply src/channel.rs:96]: No? I mean to check that one doesn't set a channel that already has been set before automatically or manually.

File: src/sctp/mod.rs (hunk 1 above)
Comment (algesten) [reply src/sctp/mod.rs:128]: Done

File: src/channel.rs (hunk 2 above)
Comment (algesten) [reply src/channel.rs]: You're right! Fixed.

**Review by k0nserv — APPROVED** [review]
Some minor things

File: src/change/direct.rs [hunk 3]
```diff
@@ -37,8 +39,8 @@ impl<'a> DirectApi<'a> {
     ///
     /// The ICE credentials consist of the username and password used by the ICE agent during
     /// the ICE session to authenticate and exchange connectivity checks with the remote peer.
-    pub fn local_ice_credentials(&self) -> &IceCreds {
```
[the remark this answers is not in the document]
Comment (algesten) [reply src/change/direct.rs:40]: True that. I need to revise this API overall.

File: src/sctp/mod.rs [hunk 4]
```diff
@@ -72,22 +72,93 @@ impl RtcSctpState {
...
+/// Reliability setting of a data channel.
+#[derive(Debug, Default, Clone, Copy, PartialEq, Eq, Hash)]
+pub enum Reliability {
+    /// Packets are delivered in order, with retransmits.
+    #[default]
+    Reliable,
+    /// Packets delivered out of order with a max lifetime.
+    MaxPacketLifetime {
+        /// The lifetime of a packet in milliseconds.
+        lifetime: u16,
```
[the remark this answers is not in the document]
Comment (algesten) [reply src/sctp/mod.rs:153]: It's the same as this: https://developer.mozilla.org/en-US/docs/Web/API/RTCPeerConnection/createDataChannel#maxpacketlifetime

Maybe should link to that doc?

[the remark this answers is not in the document]
Comment (algesten) [reply src/sctp/mod.rs:153]: But isn't a Duration is too permissive? The value, on SCTP level, is a u16 milliseconds. With Duration we can set "years". By using this datatype, we only permit the values that are actually accepted.

[the remark this answers is not in the document]
Comment (k0nserv) [reply src/sctp/mod.rs:153]: The spec already talks about the application potentially limiting it further, I guess if we want to use `u16` to enforce the 65536ms maximum and not restrict it further then we can keep it as `u16`. If we want to impose further limits we'd need custom code regardless of the type we pick

**Pull request merged**